File: 200.number-of-islands.java
/*
 * @lc app=leetcode id=200 lang=java
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) return 0;
        int res = 0;
        int[][] moves = new int[][] {{-1,0},{0,1},{1,0},{0,-1}};
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; ++i) {
            for (int j = 0; j < grid[0].length; ++j) {
                if (visited[i][j] || grid[i][j] != '1') continue;
                res++;
                dfs(visited, i, j, moves, grid);
            }
        }
        return res;
    }

    private void dfs(boolean[][] visited, int i, int j, int[][] moves, char[][] grid) {
        visited[i][j] = true;
        for (int[] move : moves) {
            int ni = i+move[0];
            int nj = j+move[1];
            if (0 <= ni && ni < grid.length && 
                0 <= nj && nj < grid[0].length && 
                grid[ni][nj] == '1' && !visited[ni][nj]) {
                dfs(visited, ni, nj, moves, grid);
            }
        }
    }
}
// @lc code=end

File: go/200.number-of-islands.go
/*
 * @lc app=leetcode id=200 lang=golang
 *
 * [200] Number of Islands
 */

// @lc code=start
func numIslands(grid [][]byte) int {
	num := 0
	visited := make([][]bool, len(grid))
	for i := 0; i < len(grid); i++ {
		visited[i] = make([]bool, len(grid[0]))
	}
	var dfs func(int, int)
	dfs = func(i, j int) {
		invalid := i < 0 || i >= len(grid)
		invalid = invalid || j < 0 || j >= len(grid[0])
		invalid = invalid || grid[i][j] == '0' || visited[i][j] == true
		if invalid {
			return
		}
		visited[i][j] = true
		dfs(i-1, j)
		dfs(i+1, j)
		dfs(i, j-1)
		dfs(i, j+1)
	}

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if !visited[i][j] && grid[i][j] == '1' {
				dfs(i, j)
				num++
			}
		}
	}
	return num
}
// @lc code=end